File: Creational/Builder/CSharp/Car.cs
// Car.cs
using System;

public class Car {
  public string Engine { get; private set; }
  public bool HasGPS { get; private set; }
  public bool HasSunroof { get; private set; }
  public string Seats { get; private set; }

  private Car(CarBuilder builder) {
    Engine = builder.Engine ?? "Unknown Engine";
    HasGPS = builder.HasGPS;
    HasSunroof = builder.HasSunroof;
    Seats = builder.Seats ?? "Standard";
  }

  public void ShowSpecifications() {
    Console.WriteLine("Car Specifications:");
    Console.WriteLine($"Engine: {Engine}");
    Console.WriteLine($"GPS: {(HasGPS ? "Yes" : "No")}");
    Console.WriteLine($"Sunroof: {(HasSunroof ? "Yes" : "No")}");
    Console.WriteLine($"Seats: {Seats}");
  }

  // Builder Class
  public class CarBuilder {
    public string? Engine { get; private set; }
    public bool HasGPS { get; private set; }
    public bool HasSunroof { get; private set; }
    public string? Seats { get; private set; }

    public CarBuilder SetEngine(string engine) {
      Engine = engine;
      return this;
    }

    public CarBuilder SetGPS(bool hasGPS) {
      HasGPS = hasGPS;
      return this;
    }

    public CarBuilder SetSunroof(bool hasSunroof) {
      HasSunroof = hasSunroof;
      return this;
    }

    public CarBuilder SetSeats(string seats) {
      Seats = seats;
      return this;
    }

    public Car Build() {
      return new Car(this);
    }
  }
}

File: Creational/Builder/Java/Car.java
// Car.java
public class Car {
  private final String engine;
  private final boolean hasGPS;
  private final boolean hasSunroof;
  private final String seats;

  private Car(CarBuilder builder) {
    this.engine = builder.engine;
    this.hasGPS = builder.hasGPS;
    this.hasSunroof = builder.hasSunroof;
    this.seats = builder.seats;
  }

  public void showSpecifications() {
    System.out.println("Car Specifications:");
    System.out.println("Engine: " + engine);
    System.out.println("GPS: " + (hasGPS ? "Yes" : "No"));
    System.out.println("Sunroof: " + (hasSunroof ? "Yes" : "No"));
    System.out.println("Seats: " + seats);
  }

  // Builder Class
  public static class CarBuilder {
    private String engine;
    private boolean hasGPS;
    private boolean hasSunroof;
    private String seats;

    public CarBuilder setEngine(String engine) {
      this.engine = engine;
      return this;
    }

    public CarBuilder setGPS(boolean hasGPS) {
      this.hasGPS = hasGPS;
      return this;
    }

    public CarBuilder setSunroof(boolean hasSunroof) {
      this.hasSunroof = hasSunroof;
      return this;
    }

    public CarBuilder setSeats(String seats) {
      this.seats = seats;
      return this;
    }

    public Car build() {
      return new Car(this);
    }
  }
}
